File: haskell/src/Day01b.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns -Wunused-top-binds #-}
{-# OPTIONS_GHC -Wno-unused-top-binds #-}

module Day01b (run) where

import AOCHelper
import Data.Char (digitToInt, isDigit)
import Data.List (isPrefixOf)
import Data.Maybe ( fromMaybe, fromJust, catMaybes, mapMaybe )
import Data.Text (Text)
import qualified Data.Text as T
-- import Data.List

digwords :: [Text]
digwords = map T.pack ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

fromDigiwords :: [Text] -> Int -> Text -> Maybe Int
fromDigiwords [] _ _ = Nothing
fromDigiwords (x : xs) n s
        | x `T.isPrefixOf` s = Just n
        | otherwise = fromDigiwords xs (n + 1) s

fdd :: Bool -> Text -> Maybe Int
fdd cw t | t == T.empty = Nothing
         | isDigit x = Just . digitToInt $ x
         | cw = fromDigiwords digwords 1 t
         | otherwise = Nothing

    where (x,_) =fromJust (T.uncons t)


fd :: Bool -> Text -> Int
fd cw s = (head . mapMaybe (fdd cw) $ T.tails s )* 10 + (head . mapMaybe (fdd cw) $ reverse (T.tails s) )

run :: IO ()
run = do
    putStrLn "--- Day 1: Trebuchet?! ---"
    putStr " Part1: "
    readInpT "day01.txt" >>= part1 >>= assertInt 55386
    putStr " Part2: "
    readInpT "day01.txt" >>= part2 >>= assertInt 54824

part1 :: Text -> IO Int
part1 s = do
    return $ sum $ map (fd False) $ T.lines s

part2 :: Text -> IO Int
part2 s = do
    return $ sum $ map (fd True) $ T.lines s

File: rust/src/day01b.rs
use itertools::Itertools;

use crate::util;
pub const NAME: &str = "Day 1: Trebuchet?! ";

pub fn data() -> String {
    util::get_input("day01.txt")
}

const W2: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn gd1(s: &str) -> Option<usize> {
    return match s[0..1].parse::<usize>() {
        Ok(n) => Some(n),
        _ => None,
    };
}

fn gd2(s: &str) -> Option<usize> {
    if let Some(n) = gd1(s) {
        return Some(n);
    }
    (1..10).filter(|i|s.starts_with(W2[i-1])).next()
}
fn slices(s: &str) -> impl Iterator<Item = &str> {
    (0..s.len()).map(|i| &s[i..])
}

fn firslastprod<'a, I>(mut q: I) -> usize
where
    I: Iterator<Item = usize>,
{
    let f = q.next().unwrap();
    match q.last() {
        Some(v) => f * 10 + v,
        None => f + f * 10,
    }
}
// 55386
pub fn part01(contents: String) -> usize {
    let qq: usize = contents
        .lines()
        .map(|s| firslastprod(slices(s).map(gd1).flatten()))
        .sum();
    qq
}

/*
54824
*/
pub fn part02(contents: String) -> usize {
    let qq: usize = contents
        .lines()
        .map(|s| firslastprod(slices(s).map(gd2).flatten()))
        .sum();
    qq
}
#[cfg(test)]
mod test_result {
    use super::*;
    #[test]
    fn part01() {
        assert_eq!(super::part01(data()), 55386);
    }
    #[test]
    fn part02() {
        assert_eq!(super::part02(data()), 54824);
    }
}

// other ppls solutions
mod others {
    #![allow(unused)]

    use crate::util;

    pub fn part01b() {
        let contents: String = util::get_input("day01.txt");
        println!("{}", parse_input(contents.as_str(), false))
    }

    pub fn part02b() {
        let contents: String = util::get_input("day01.txt");
        println!("{}", parse_input(contents.as_str(), true))
    }

    pub fn part01c() {
        let contents: String = util::get_input("day01.txt");
        println!("{}", part_one(contents.as_str()))
    }

    pub fn part02c() {
        let contents: String = util::get_input("day01.txt");
        println!("{}", part_two(contents.as_str()))
    }

    fn parse_input(input: &str, replace: bool) -> u32 {
        input
            .lines()
            .filter(|line| !line.is_empty())
            .map(|line| {
                if replace {
                    line.to_string()
                        .replace("one", "one1one")
                        .replace("two", "two2two")
                        .replace("three", "three3three")
                        .replace("four", "four4four")
                        .replace("five", "five5five")
                        .replace("six", "six6six")
                        .replace("seven", "seven7seven")
                        .replace("eight", "eight8eight")
                        .replace("nine", "nine9nine")
                } else {
                    line.to_string()
                }
            })
            .map(|line| {
                line.chars()
                    .filter_map(|c| c.to_digit(10))
                    .collect::<Vec<u32>>()
            })
            .map(|vec| 10 * vec.first().unwrap() + vec.last().unwrap())
            .sum()
    }

    //
    const LUT: [&str; 9] = [
        "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
    ];

    pub fn part_one(input: &str) -> u32 {
        input.lines().map(parse_line_1).sum()
    }

    pub fn part_two(input: &str) -> u32 {
        input.lines().map(parse_line_2).sum()
    }

    fn parse_line_1(line: &str) -> u32 {
        let first = line.chars().find_map(|c| c.to_digit(10));
        let last = line.chars().rev().find_map(|c| c.to_digit(10));
        10 * first.unwrap() + last.unwrap()
    }

    fn parse_line_2(line: &str) -> u32 {
        let first = find_pattern(0..line.len(), line);
        let last = find_pattern((0..line.len()).rev(), line);
        10 * first + last
    }

    fn find_pattern(mut it: impl Iterator<Item = usize>, line: &str) -> u32 {
        it.find_map(|i| compare_slice(&line[i..])).unwrap()
    }

    fn compare_slice(slice: &str) -> Option<u32> {
        LUT.iter()
            .enumerate()
            .find(|(_, pattern)| slice.starts_with(*pattern))
            .map(|(i, _)| i as u32 + 1)
            .or_else(|| slice.chars().next().unwrap().to_digit(10))
    }
}
